main/src/mill/modules/Assembly.scala

```diff
@@ -27,9 +27,28 @@ object Assembly {
     case class Append(path: String, separator: String = defaultSeparator) extends Rule
 
     object AppendPattern {
-      def apply(pattern: String): AppendPattern = AppendPattern(Pattern.compile(pattern))
+      def apply(pattern: Pattern): AppendPattern = new AppendPattern(pattern, defaultSeparator)
+      def apply(pattern: String): AppendPattern = apply(pattern, defaultSeparator)
+      def apply(pattern: String, separator: String): AppendPattern =
+        new AppendPattern(Pattern.compile(pattern), separator)
+
+      @deprecated(message = "Binary compatibility shim. To be removed", since = "mill 0.10.1")
+      def unapply(value: AppendPattern): Option[Pattern] = Some(value.pattern)
+    }
+    class AppendPattern(val pattern: Pattern, val separator: String) extends Rule {
+      def this(pattern: Pattern) = this(pattern, defaultSeparator)
+
+      override def productArity: Int = 2
```

Comment:
This should be deprecated with the same message as below.

Reply:
Why? It's not a shim, valid and usable method, moreover rule extends product.

Reply:
> moreover rule extends product.

Then you're right. We can deprecate the product things later for `Rule` and its subclasses.